### ISSUE TITLE
Smoothing on Mean Encoder

### ISSUE THREAD

**Issue by hectorpatino** [feature-engine/feature_engine#315]
**Is your feature request related to a problem? Please describe.**
Hi Sole, currently Im doing the mini course of [feature engineering](https://www.kaggle.com/ryanholbrook/target-encoding) on kaggle. The last seccion is about Mean Encoding and how it can create risks of overfitting specially with unknown or rare categories. They mention that smooting is a way to deal with this. I checked the [code ](https://github.com/feature-engine/feature_engine/blob/main/feature_engine/encoding/mean_encoding.py#L12) but I don't see anything like that. So ¿should MeanEncoder has a smoothing parameter?

**Comment by solegalli** [comment]
Yes, some users also requested to introduce this transformer or smoothing function in this issue: #160 

I am still wondering if it is worth to bring in to Feature-engine the functionality that already exists in category_encoders. Category encoders also allows you to select columns, so not sure how much value it would add.

What are your thoughts?

**Comment by hectorpatino** [comment]
As someone mention in the issue there is no need to reinvent the wheel. Maybe an update on docstring suggesting `category_encoders` for smoothing?

**Comment by solegalli** [comment]
Great! In that case, please make a PR to the branch "pydata_template" as the doc files have been reorganized quite a bit compared to master. 

And you probably want to modify [this file](https://github.com/feature-engine/feature_engine/blob/pydata_template/docs/user_guide/encoding/MeanEncoder.rst) and / or [this file](https://github.com/feature-engine/feature_engine/blob/pydata_template/feature_engine/encoding/mean_encoding.py).

**Comment by solegalli** [comment]
Already added links to category encoders in the class api docs. We still need to add some details in the user guide.

**Comment by hectorpatino** [comment]
Ive been extremely busy sole but I Will try to do it this weekend

**Comment by solegalli** [comment]
added links to docs in #314

**Comment by solegalli** [comment]
Reopening as I think it might be worth bringing in this functionality to our class as well. It is not so complicated.

**Comment by Morgan-Sell** [comment]
hola @solegalli,

I can work on this task. Although, I'm always wary when you write, "It is not so complicated." ;)

**Comment by solegalli** [comment]
Go for it. 

We need to implement the transformation described in the article cited in [this documentation](https://contrib.scikit-learn.org/category_encoders/targetencoder.html) and I guess we can borrow the logic to calculate the prior and posterior probability from category encoders as well.

What I am still not sure, is if we should adapt our MeanEncoder to be able to apply the smoothing, or create a complete new class. The reason for the above is that the MeanEncoder is used by other transformers. Although these other transformers could also benefit from the smoothing. We need to brainstorm a bit.

**Comment by GLevV** [comment]
If we want to distinguish our mean encoder from many others (category_encoders, h2o, dirty_cat etc) we could use different smoothing function. In original Micci-Barreca they are using sigmoid function to calculate smoothing factors, but it possible to use any other monotonic (0-1) function (they also said it in the paper). For example, [here](https://tel.archives-ouvertes.fr/tel-02614322/document) authors suggest using x / (x + a) (p. 23, formula 2.14).

**Comment by solegalli** [comment]
thank you @GLevV

**Comment by Morgan-Sell** [comment]
Ok, I'm getting back to this issue. I'll get started on it this coming week.

@GLevV, a few questions:

1. Are you referring to using x / (x + a) for the weighting? 
2. Does `x` representhe total number of times that category occurs in the data? 
3. Does `m` represent the "smoothing factor" that is to be selected by the user?

**Comment by GLevV** [comment]
x are the counts/frequencies of categories, a is a smoothing factor. If a is 0 then this become vanilla MeanEncoder as we have right now (which is good for backward compatibility and classes relying on MeanEncoder). 
For more math details please refer to the papers mentioned above.

On the footnote, in Micci-Barecca paper they mentioned that a for this formula could be easily calculated from the data. This is interesting, we can try to 'auto' option for a.

**Comment by Morgan-Sell** [comment]
Ok, cool. Thanks for the explanation. We're saying the same thing.

@solegalli, should we knock out SelectByInformationValue and TargetMeanSelector before jumping into this one? I see it's a priority.

**Comment by GLevV** [comment]
Is it in the workings or stale?

**Comment by solegalli** [comment]
It is waiting for a kind contributor to pick it up :)
